### ISSUE TITLE
AOT Autograd - Functionalization pass with torch.dispatch

### ISSUE THREAD

**Issue by anijain2305** [pytorch/torchdynamo#88]
Run functionalization to resolve mutation related errors in AOT Autograd.

**Comment by anijain2305** [comment]
WIP PR in functorch - https://github.com/pytorch/functorch/pull/703

cc @bdhirsh 

Issues seen in functionalization by running it on Torchbenhc
- [ ] zero_
- [ ] index
- [ ] _fused_moving_avg_obs_fq_helper

**Comment by ciciewang** [comment]
@bdhirsh is working on it right now.

**Comment by anijain2305** [comment]
New issues


- [ ] index

The repro is as follows
~~~
import torch
from torch.nn import *
from functorch.experimental import functionalize

class FxModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer('_tensor_constant0', torch.zeros([4, 1], dtype=torch.int64))
        self.register_buffer('_tensor_constant1', torch.zeros([4, 359613], dtype=torch.int64))

    
    
    def forward(self, primals_1, primals_2):
        view = torch.ops.aten.view(primals_2, [4, 359613]);  primals_2 = None
        split_with_sizes = torch.ops.aten.split_with_sizes(view, [269952, 67488, 16872, 4218, 1083], 1)
        getitem = split_with_sizes[0]
        getitem_1 = split_with_sizes[1]
        getitem_2 = split_with_sizes[2]
        getitem_3 = split_with_sizes[3]
        getitem_4 = split_with_sizes[4];  split_with_sizes = None
        topk = torch.ops.aten.topk(getitem, 1000, 1);  getitem = None
        getitem_5 = topk[0]
        getitem_6 = topk[1];  topk = None
        add = torch.ops.aten.add(getitem_6, 0);  getitem_6 = None
        topk_1 = torch.ops.aten.topk(getitem_1, 1000, 1);  getitem_1 = None
        getitem_7 = topk_1[0]
        getitem_8 = topk_1[1];  topk_1 = None
        add_1 = torch.ops.aten.add(getitem_8, 269952);  getitem_8 = None
        topk_2 = torch.ops.aten.topk(getitem_2, 1000, 1);  getitem_2 = None
        getitem_9 = topk_2[0]
        getitem_10 = topk_2[1];  topk_2 = None
        add_2 = torch.ops.aten.add(getitem_10, 337440);  getitem_10 = None
        topk_3 = torch.ops.aten.topk(getitem_3, 1000, 1);  getitem_3 = None
        getitem_11 = topk_3[0]
        getitem_12 = topk_3[1];  topk_3 = None
        add_3 = torch.ops.aten.add(getitem_12, 354312);  getitem_12 = None
        topk_4 = torch.ops.aten.topk(getitem_4, 1000, 1);  getitem_4 = None
        getitem_13 = topk_4[0]
        getitem_14 = topk_4[1];  topk_4 = None
        add_4 = torch.ops.aten.add(getitem_14, 358530);  getitem_14 = None
        cat = torch.ops.aten.cat([add, add_1, add_2, add_3, add_4], 1);  add = add_1 = add_2 = add_3 = add_4 = None
        _tensor_constant0 = self._tensor_constant0
        index = torch.ops.aten.index(view, [_tensor_constant0, cat]);  view = _tensor_constant0 = None
        _tensor_constant1 = self._tensor_constant1
        _tensor_constant0_1 = self._tensor_constant0
        index_1 = torch.ops.aten.index(_tensor_constant1, [_tensor_constant0_1, cat]);  _tensor_constant1 = _tensor_constant0_1 = None
        _tensor_constant0_2 = self._tensor_constant0
        index_2 = torch.ops.aten.index(primals_1, [_tensor_constant0_2, cat]);  primals_1 = _tensor_constant0_2 = cat = None
        sigmoid = torch.ops.aten.sigmoid(index);  index = None
        select = torch.ops.aten.select(index_2, 0, 0)
        select_1 = torch.ops.aten.select(index_2, 0, 1)
        select_2 = torch.ops.aten.select(index_2, 0, 2)
        select_3 = torch.ops.aten.select(index_2, 0, 3);  index_2 = None
        select_4 = torch.ops.aten.select(sigmoid, 0, 0)
        select_5 = torch.ops.aten.select(sigmoid, 0, 1)
        select_6 = torch.ops.aten.select(sigmoid, 0, 2)
        select_7 = torch.ops.aten.select(sigmoid, 0, 3);  sigmoid = None
        select_8 = torch.ops.aten.select(index_1, 0, 0)
        select_9 = torch.ops.aten.select(index_1, 0, 1)
        select_10 = torch.ops.aten.select(index_1, 0, 2)
        select_11 = torch.ops.aten.select(index_1, 0, 3);  index_1 = None
        slice_1 = torch.ops.aten.slice(select, 1, 0, 9223372036854775807, 2)
        slice_2 = torch.ops.aten.slice(select, 1, 1, 9223372036854775807, 2);  select = None
        clamp = torch.ops.aten.clamp(slice_1, 0, 1199);  slice_1 = None
        clamp_1 = torch.ops.aten.clamp(slice_2, 0, 799);  slice_2 = None
        stack = torch.ops.aten.stack([clamp, clamp_1], 2);  clamp = clamp_1 = None
        view_1 = torch.ops.aten.view(stack, [5000, 4]);  stack = None
        return [select_1, select_5, select_9, select_2, select_6, select_10, select_3, select_7, select_11, view_1, select_4, select_8, None, None]
        

primals_sizes = [torch.Size([4, 359613, 4]), torch.Size([1438452, 1])]
primals_dtypes = [torch.float32, torch.float32]



mod = FxModule().to(device="cpu")
primals = [torch.empty(size, dtype=dtype, device="cpu") for (size, dtype) in zip(primals_sizes, primals_dtypes)]

inputs = primals
print(mod(*inputs))



res = functionalize(mod)(*inputs)
print(res)
~~~

 
- [ ] Quantized models
failure log - https://www.internalfb.com/phabricator/paste/view/P503611308


- [ ] Constants not on GPU device - *Not* a Functionalize problem - https://github.com/pytorch/functorch/issues/815

**Comment by ezyang** [comment]
@bdhirsh index still triggers assert error

```
RuntimeError: functional_count == 0 || nonfunctional_count == 0 INTERNAL ASSERT FAILED at "/raid/ezyang/pytorch-scratch2/aten/src/ATe
n/FunctionalTensorWrapper.cpp":549, please report a bug to PyTorch. Functionalization encountered a list of tensors where some are fu
nctionaland some are not, which is not currently unsupported.   
```

**Comment by anijain2305** [comment]
Closed in favor of #789
